File: bridge/src/registry/materials.ts
import { z } from "zod";
import { callArborJson } from "../ue5-client.js";
import type { CategoryTool } from "./types.js";

export const materialsTool: CategoryTool = {
  description:
    "Material creation, assignment, and graph editing: create simple/textured/PBR/world-aligned " +
    "materials, create material instances with parameter overrides, assign materials to actors, " +
    "and edit material graphs via granular primitives (add/remove/connect expressions) or the " +
    "build_material orchestrator that takes a full JSON spec.",

  readOnlyActions: ["query", "list_expression_types", "get_expression_class_params", "query_function"],

  actionParams: {
    create: {
      summary: "Create simple PBR material with color and properties",
      optional: ["name", "content_path", "color", "metallic", "roughness", "opacity", "emissive_color", "emissive_strength", "two_sided"],
    },
    create_from_textures: {
      summary: "Create material from texture maps (albedo, normal, roughness, etc.)",
      optional: ["name", "content_path", "albedo", "normal", "roughness_map", "metallic_map", "ao", "height_map", "tiling"],
    },
    create_pbr: {
      summary: "Create parameterized PBR material with adjustable properties",
      optional: ["name", "content_path", "color", "metallic", "roughness", "opacity", "emissive_color", "emissive_strength", "two_sided", "tiling"],
    },
    create_instance: {
      summary: "Create material instance with parameter overrides",
      required: ["parent_material"],
      optional: ["name", "content_path", "scalar_params", "vector_params", "texture_params"],
    },
    create_world_aligned: {
      summary: "Create world-aligned material for seamless tiling across surfaces",
      optional: ["name", "content_path", "color", "metallic", "roughness", "tiling"],
    },
    assign: {
      summary: "Assign material to actors",
      required: ["actor_names", "material_path"],
      optional: ["slot_index"],
    },
    ensure_pbr_base: {
      summary: "Ensure PBR base material exists in content path",
      optional: ["content_path"],
    },
    query: {
      summary: "Dump a material's expression graph as JSON (expressions, connections, flags)",
      required: ["material_path"],
    },
    list_expression_types: {
      summary: "List UMaterialExpression subclasses, optional substring filter",
      optional: ["filter"],
    },
    get_expression_class_params: {
      summary: "Reflect on an expression class for its editable properties + defaults",
      required: ["expression_class"],
    },
    add_expression: {
      summary: "Add an expression to a material's graph; returns the sentinel ID",
      required: ["material_path", "expression_class"],
      optional: ["expression_id", "properties", "node_x", "node_y"],
    },
    remove_expression: {
      summary: "Remove an expression by sentinel ID",
      required: ["material_path", "expression_id"],
    },
    set_expression_property: {
      summary: "Set a single property by name on an existing expression",
      required: ["material_path", "expression_id", "property_name", "value"],
    },
    connect_nodes: {
      summary: "Wire one expression's output to another expression's input",
      required: ["material_path", "from_id", "to_id"],
      optional: ["from_output", "to_input"],
    },
    connect_output: {
      summary: "Wire an expression's output to a material output channel (BaseColor, Normal, ..., or FrontMaterial on 5.7+)",
      required: ["material_path", "expression_id", "property"],
      optional: ["from_output"],
    },
    recompile: {
      summary: "Explicit terminal recompile + save after batched edits",
      required: ["material_path"],
    },
    build: {
      summary: "Build or update a complete material from a JSON spec (idempotent). See build_material docs for schema.",
      required: ["spec"],
    },
    query_function: {
      summary: "Dump a Material Function's graph as JSON (expressions, connections, inputs, outputs)",
      required: ["function_path"],
    },
    build_function: {
      summary: "Build or update a complete Material Function from a JSON spec (idempotent). No outputs/flags block; inputs/outputs are FunctionInput/FunctionOutput nodes. See BuildMaterialFunction docs for schema.",
      required: ["spec"],
    },
    layout: {
      summary: "Auto-arrange a material or material function's nodes into a readable left-to-right layout (UE built-in). build/build_function already do this unless the spec sets auto_layout:false.",
      required: ["path"],
    },
  },

  schema: {
    name: z.string().optional().describe("Material name"),
    content_path: z.string().optional().describe("Content folder. Default: /Game/Materials"),
    // create (simple)
    color: z.object({ r: z.number(), g: z.number(), b: z.number() }).optional().describe("Base color {r,g,b} 0-1"),
    metallic: z.number().optional().describe("Metallic value 0-1"),
    roughness: z.number().optional().describe("Roughness value 0-1"),
    opacity: z.number().optional().describe("Opacity 0-1 (enables translucency)"),
    emissive_color: z.object({ r: z.number(), g: z.number(), b: z.number() }).optional().describe("Emissive color"),
    emissive_strength: z.number().optional().describe("Emissive multiplier"),
    two_sided: z.boolean().optional().describe("Two-sided material"),
    // create_from_textures
    albedo: z.string().optional().describe("Albedo texture path"),
    normal: z.string().optional().describe("Normal map texture path"),
    roughness_map: z.string().optional().describe("Roughness texture path"),
    metallic_map: z.string().optional().describe("Metallic texture path"),
    ao: z.string().optional().describe("AO texture path"),
    height_map: z.string().optional().describe("Height map texture path"),
    tiling: z.number().optional().describe("UV tiling factor"),
    // create_instance
    parent_material: z.string().optional().describe("Parent material path (create_instance)"),
    scalar_params: z.record(z.number()).optional().describe("Scalar param overrides"),
    vector_params: z.record(z.object({ r: z.number(), g: z.number(), b: z.number(), a: z.number().optional() })).optional()
      .describe("Vector param overrides"),
    texture_params: z.record(z.string()).optional().describe("Texture param overrides (name → texture path)"),
    // assign
    actor_names: z.array(z.string()).optional().describe("Actor names to assign material to"),
    material_path: z.string().optional().describe("Material content path to assign"),
    slot_index: z.number().optional().describe("Material slot index. Default 0"),
    // graph editing
    filter: z.string().optional().describe("Substring filter for list_expression_types"),
    expression_class: z.string().optional().describe("UMaterialExpression subclass name (U-prefix optional)"),
    expression_id: z.string().optional().describe("Stable Arbor ID stamped into the expression's Desc field"),
    properties: z.record(z.any()).optional().describe("Property name -> value dict; FLinearColor as [r,g,b,a], textures as asset path string, enums as the enum value name"),
    node_x: z.number().optional().describe("Node X position in graph"),
    node_y: z.number().optional().describe("Node Y position in graph"),
    property_name: z.string().optional().describe("Property name for set_expression_property"),
    value: z.any().optional().describe("Property value for set_expression_property"),
    from_id: z.string().optional().describe("Source expression ID for connect_nodes"),
    to_id: z.string().optional().describe("Target expression ID for connect_nodes"),
    from_output: z.string().optional().describe("Source output pin name (empty = first)"),
    to_input: z.string().optional().describe("Target input pin name (empty = first; warning if target has >1 inputs)"),
    property: z.string().optional().describe("Material output property: BaseColor/Normal/Roughness/Metallic/EmissiveColor/Opacity/AmbientOcclusion/WorldPositionOffset"),
    spec: z.record(z.any()).optional().describe("Full BuildMaterial / BuildMaterialFunction spec — see ArborMaterialGraphTools.h for schema"),
    function_path: z.string().optional().describe("Material Function asset path for query_function (e.g. /Game/Materials/Functions/MF_SDF_Circle)"),
    path: z.string().optional().describe("Material or Material Function asset path for layout (auto-detects type)"),
  },

  actions: {
    async create(p) {
      return callArborJson("ArborMaterialTools", "CreateMaterial", {
        ParamsJson: JSON.stringify({
          name: p.name, content_path: p.content_path,
          color: p.color, metallic: p.metallic, roughness: p.roughness,
          opacity: p.opacity, emissive_color: p.emissive_color,
          emissive_strength: p.emissive_strength, two_sided: p.two_sided,
        }),
      });
    },

    async create_from_textures(p) {
      return callArborJson("ArborMaterialTools", "CreateMaterialFromTextures", {
        ParamsJson: JSON.stringify({
          name: p.name, content_path: p.content_path,
          albedo: p.albedo, normal: p.normal,
          roughness_map: p.roughness_map, metallic_map: p.metallic_map,
          ao: p.ao, height_map: p.height_map, tiling: p.tiling,
        }),
      });
    },

    async create_pbr(p) {
      return callArborJson("ArborMaterialTools", "CreateParameterizedPBRMaterial", {
        ParamsJson: JSON.stringify({
          name: p.name, content_path: p.content_path,
          color: p.color, metallic: p.metallic, roughness: p.roughness,
          opacity: p.opacity, emissive_color: p.emissive_color,
          emissive_strength: p.emissive_strength, two_sided: p.two_sided,
          tiling: p.tiling,
        }),
      });
    },

    async create_instance(p) {
      if (!p.parent_material) throw new Error("parent_material required");
      return callArborJson("ArborMaterialTools", "CreateMaterialInstance", {
        ParamsJson: JSON.stringify({
          name: p.name, content_path: p.content_path,
          parent_material: p.parent_material,
          scalar_params: p.scalar_params, vector_params: p.vector_params,
          texture_params: p.texture_params,
        }),
      });
    },

    async create_world_aligned(p) {
      return callArborJson("ArborMaterialTools", "CreateWorldAlignedMaterial", {
        ParamsJson: JSON.stringify({
          name: p.name, content_path: p.content_path,
          color: p.color, metallic: p.metallic, roughness: p.roughness,
          tiling: p.tiling,
        }),
      });
    },

    async assign(p) {
      if (!p.actor_names || !p.material_path) throw new Error("actor_names, material_path required");
      return callArborJson("ArborMaterialTools", "AssignMaterial", {
        ParamsJson: JSON.stringify({
          actor_names: p.actor_names, material_path: p.material_path,
          slot_index: p.slot_index ?? 0,
        }),
      });
    },

    async ensure_pbr_base(p) {
      return callArborJson("ArborMaterialTools", "EnsurePBRBaseMaterial", {
        ContentPath: (p.content_path as string) ?? "/Game/Materials",
      });
    },

    // ---- Graph editing primitives (Phase 1) ----

    async query(p) {
      if (!p.material_path) throw new Error("material_path required");
      return callArborJson("ArborMaterialGraphTools", "QueryMaterial", {
        MaterialPath: p.material_path as string,
      });
    },

    async list_expression_types(p) {
      return callArborJson("ArborMaterialGraphTools", "ListMaterialExpressionTypes", {
        Filter: (p.filter as string) ?? "",
      });
    },

    async get_expression_class_params(p) {
      if (!p.expression_class) throw new Error("expression_class required");
      return callArborJson("ArborMaterialGraphTools", "GetMaterialExpressionClassParams", {
        ClassName: p.expression_class as string,
      });
    },

    async add_expression(p) {
      if (!p.material_path || !p.expression_class) throw new Error("material_path, expression_class required");
      return callArborJson("ArborMaterialGraphTools", "AddMaterialExpression", {
        ParamsJson: JSON.stringify({
          material_path: p.material_path,
          expression_class: p.expression_class,
          expression_id: p.expression_id,
          properties: p.properties,
          node_x: p.node_x ?? 0,
          node_y: p.node_y ?? 0,
        }),
      });
    },

    async remove_expression(p) {
      if (!p.material_path || !p.expression_id) throw new Error("material_path, expression_id required");
      return callArborJson("ArborMaterialGraphTools", "RemoveMaterialExpressionById", {
        MaterialPath: p.material_path as string,
        ExpressionId: p.expression_id as string,
      });
    },

    async set_expression_property(p) {
      if (!p.material_path || !p.expression_id || !p.property_name)
        throw new Error("material_path, expression_id, property_name required");
      return callArborJson("ArborMaterialGraphTools", "SetMaterialExpressionProperty", {
        ParamsJson: JSON.stringify({
          material_path: p.material_path,
          expression_id: p.expression_id,
          property_name: p.property_name,
          value: p.value,
        }),
      });
    },

    async connect_nodes(p) {
      if (!p.material_path || !p.from_id || !p.to_id)
        throw new Error("material_path, from_id, to_id required");
      return callArborJson("ArborMaterialGraphTools", "ConnectMaterialNodes", {
        ParamsJson: JSON.stringify({
          material_path: p.material_path,
          from_id: p.from_id, to_id: p.to_id,
          from_output: p.from_output ?? "",
          to_input: p.to_input ?? "",
        }),
      });
    },

    async connect_output(p) {
      if (!p.material_path || !p.expression_id || !p.property)
        throw new Error("material_path, expression_id, property required");
      return callArborJson("ArborMaterialGraphTools", "ConnectMaterialOutput", {
        ParamsJson: JSON.stringify({
          material_path: p.material_path,
          expression_id: p.expression_id,
          property: p.property,
          from_output: p.from_output ?? "",
        }),
      });
    },

    async recompile(p) {
      if (!p.material_path) throw new Error("material_path required");
      return callArborJson("ArborMaterialGraphTools", "RecompileMaterialAsset", {
        MaterialPath: p.material_path as string,
      });
    },

    // ---- BuildMaterial orchestrator (Phase 2) ----

    async build(p) {
      if (!p.spec) throw new Error("spec required");
      return callArborJson("ArborMaterialGraphTools", "BuildMaterial", {
        JsonSpec: JSON.stringify(p.spec),
      });
    },

    // ---- Material Function authoring (Phase 7) ----

    async query_function(p) {
      if (!p.function_path) throw new Error("function_path required");
      return callArborJson("ArborMaterialGraphTools", "QueryMaterialFunction", {
        FunctionPath: p.function_path as string,
      });
    },

    async build_function(p) {
      if (!p.spec) throw new Error("spec required");
      return callArborJson("ArborMaterialGraphTools", "BuildMaterialFunction", {
        JsonSpec: JSON.stringify(p.spec),
      });
    },

    // ---- Graph layout ----

    async layout(p) {
      const path = (p.path ?? p.material_path ?? p.function_path) as string | undefined;
      if (!path) throw new Error("path required");
      return callArborJson("ArborMaterialGraphTools", "LayoutMaterial", {
        JsonParams: JSON.stringify({ path }),
      });
    },
  },
};

File: Content/Python/arbor/materials.py
"""Arbor materials — create, configure, and assign materials (C++ backend)."""

import json
import unreal

from arbor.utils import _to_linear_color, _resolve_actor, load_asset, write_result


def _call_cpp(result_json):
    """Parse C++ JSON result, log errors, return parsed dict."""
    result = json.loads(result_json)
    if result.get("error"):
        unreal.log_error(f"[arbor.materials] {result['error']}")
    return result


def _force_compile(material_path):
    """Compile a material's shaders synchronously so previews/thumbnails reflect
    the real material instead of the default (lit-grey) placeholder.

    The thumbnail + viewport pipeline shows the default material while shaders
    compile asynchronously. Forcing AsyncCompiling off + recompiling in the same
    call makes the result reliable. Best-effort; never raises.
    """
    try:
        unreal.SystemLibrary.execute_console_command(None, "r.ShaderCompiler.AsyncCompiling 0")
        asset = unreal.load_asset(material_path)
        if isinstance(asset, unreal.Material):
            unreal.MaterialEditingLibrary.recompile_material(asset)
    except Exception as e:  # noqa: BLE001 - diagnostics only
        unreal.log_warning(f"[arbor.materials] _force_compile({material_path}): {e}")


# ---------------------------------------------------------------------------
# Public API — all delegate to ArborMaterialTools C++
# ---------------------------------------------------------------------------

def create_material(name, content_path="/Game/Materials", color=(1, 1, 1),
                    metallic=0.0, roughness=0.5):
    """Create a simple material with constant base color, metallic, and roughness.

    Returns:
        The created ``unreal.Material``, or ``None``.
    """
    params = json.dumps({
        "name": name, "content_path": content_path,
        "color": list(color), "metallic": metallic, "roughness": roughness,
    })
    result = _call_cpp(unreal.ArborMaterialTools.create_material(params))
    if not result.get("success"):
        return None
    return load_asset(result.get("asset_path", ""))


def create_material_from_textures(name, content_path="/Game/Materials",
                                  base_color_path=None, normal_path=None,
                                  roughness_path=None, metallic_path=None):
    """Create a material with texture samplers wired to material outputs.

    Returns:
        The created ``unreal.Material``, or ``None``.
    """
    p = {"name": name, "content_path": content_path}
    if base_color_path: p["base_color_path"] = base_color_path
    if normal_path: p["normal_path"] = normal_path
    if roughness_path: p["roughness_path"] = roughness_path
    if metallic_path: p["metallic_path"] = metallic_path

    result = _call_cpp(unreal.ArborMaterialTools.create_material_from_textures(json.dumps(p)))
    if not result.get("success"):
        return None
    return load_asset(result.get("asset_path", ""))


def create_material_instance(parent_path, name, content_path="/Game/Materials",
                             params=None):
    """Create a MaterialInstanceConstant from a parent material.

    Returns:
        The created ``MaterialInstanceConstant``, or ``None``.
    """
    p = {"parent_path": parent_path, "name": name, "content_path": content_path}
    if params:
        # Convert tuple values to lists for JSON serialization
        serializable_params = {}
        for k, v in params.items():
            if isinstance(v, tuple):
                serializable_params[k] = list(v)
            else:
                serializable_params[k] = v
        p["params"] = serializable_params

    result = _call_cpp(unreal.ArborMaterialTools.create_material_instance(json.dumps(p)))
    if not result.get("success"):
        return None
    return load_asset(result.get("asset_path", ""))


def create_parameterized_pbr_material(name, content_path="/Game/Materials",
                                      default_tiling=1.0):
    """Create a PBR base material with parameterized textures and tiling.

    Returns:
        Content path to the created Material, or ``None``.
    """
    params = json.dumps({
        "name": name, "content_path": content_path,
        "default_tiling": default_tiling,
    })
    result = _call_cpp(unreal.ArborMaterialTools.create_parameterized_pbr_material(params))
    if not result.get("success"):
        return None
    return result.get("asset_path")


def ensure_pbr_base_material(content_path="/Game/Materials"):
    """Return the path to the shared parameterized PBR base material.

    Creates it if it doesn't already exist.

    Returns:
        Content path string.
    """
    result = _call_cpp(unreal.ArborMaterialTools.ensure_pbr_base_material(content_path))
    if not result.get("success"):
        return None
    return result.get("asset_path")


def create_world_aligned_material(name, content_path="/Game/Materials",
                                  base_color_path=None, normal_path=None,
                                  roughness_path=None, metallic_path=None,
                                  ao_path=None, tiling_scale=200.0):
    """Create a PBR material using world-aligned (tri-planar) texture projection.

    Returns:
        Content path string to the created material, or ``None``.
    """
    p = {"name": name, "content_path": content_path, "tiling_scale": tiling_scale}
    if base_color_path: p["base_color_path"] = base_color_path
    if normal_path: p["normal_path"] = normal_path
    if roughness_path: p["roughness_path"] = roughness_path
    if metallic_path: p["metallic_path"] = metallic_path
    if ao_path: p["ao_path"] = ao_path

    result = _call_cpp(unreal.ArborMaterialTools.create_world_aligned_material(json.dumps(p)))
    if not result.get("success"):
        return None
    return result.get("asset_path")


def _assign_material_single(actor, material_path, slot=0):
    """Assign a material to one actor. Returns a result dict."""
    try:
        actor = _resolve_actor(actor)
        if actor is None:
            return {"success": False, "error": "Actor not found"}

        mat = material_path if not isinstance(material_path, str) else load_asset(material_path)
        if mat is None:
            return {"success": False, "error": f"Material not found: {material_path}"}

        comp = actor.get_component_by_class(unreal.StaticMeshComponent)
        if comp is None:
            label = actor.get_actor_label()
            return {"success": False, "error": f"No StaticMeshComponent on '{label}'"}

        comp.set_material(slot, mat)
        label = actor.get_actor_label()
        return {"success": True, "actor": label, "slot": slot}
    except Exception as e:
        return {"success": False, "error": str(e)}


def assign_material(actor, material_path, slot=0):
    """Assign a material to an actor's static mesh component."""
    result = _assign_material_single(actor, material_path, slot)
    write_result(result)
    return result


def assign_material_by_name(actor_name, material_path, slot=0):
    """Find actor(s) by name and assign a material."""
    if isinstance(actor_name, (list, tuple)):
        # Batch assignment via C++
        p = {
            "actor_names": list(actor_name),
            "material_path": material_path if isinstance(material_path, str) else "",
            "slot": slot,
        }
        result_json = unreal.ArborMaterialTools.assign_material(json.dumps(p))
        result = json.loads(result_json)
    else:
        result = _assign_material_single(actor_name, material_path, slot)
    write_result(result)
    return result


# ---------------------------------------------------------------------------
# Material graph editing (Phase 1 + Phase 2) — delegates to ArborMaterialGraphTools
#
# Identify expressions by stable Arbor IDs stamped into the expression's Desc
# field. IDs survive editor save/reload. Each granular call marks the asset
# dirty but does NOT recompile — call recompile_material() once when done, or
# use build_material(spec) which batches everything in one FMaterialUpdateContext.
# ---------------------------------------------------------------------------

def query_material(path):
    """Return a JSON snapshot of a material's expression graph.

    Returns:
        dict with keys: success, expressions, connections, flags, material_path.
    """
    return _call_cpp(unreal.ArborMaterialGraphTools.query_material(path))


def list_expression_types(filter=""):
    """List all UMaterialExpression subclasses, optional substring filter."""
    return _call_cpp(unreal.ArborMaterialGraphTools.list_material_expression_types(filter or ""))


def get_expression_class_params(class_name):
    """Reflect on an expression class for its editable properties + defaults."""
    return _call_cpp(unreal.ArborMaterialGraphTools.get_material_expression_class_params(class_name))


def add_expression(material_path, expression_class, *, expression_id=None,
                   properties=None, node_x=0, node_y=0):
    """Add an expression to a material. Returns dict with success + expression_id.

    Properties values match the JSON wire format:
        scalars/strings/bools as-is; FLinearColor as [r,g,b,a] or {R,G,B,A};
        FVector as [x,y,z]; UObject refs as asset path string; enums as the
        enum value name (e.g. "SAMPLERTYPE_Normal").
    """
    p = {
        "material_path": material_path,
        "expression_class": expression_class,
        "node_x": node_x, "node_y": node_y,
    }
    if expression_id:
        p["expression_id"] = expression_id
    if properties:
        p["properties"] = properties
    return _call_cpp(unreal.ArborMaterialGraphTools.add_material_expression(json.dumps(p)))


def remove_expression(material_path, expression_id):
    """Remove an expression by its sentinel ID."""
    return _call_cpp(unreal.ArborMaterialGraphTools.remove_material_expression_by_id(
        material_path, expression_id))


def set_expression_property(material_path, expression_id, property_name, value):
    """Set a property by name on an existing expression."""
    p = {
        "material_path": material_path,
        "expression_id": expression_id,
        "property_name": property_name,
        "value": value,
    }
    return _call_cpp(unreal.ArborMaterialGraphTools.set_material_expression_property(json.dumps(p)))


def connect_nodes(material_path, from_id, to_id, *, from_output="", to_input=""):
    """Wire one expression's output to another expression's input pin."""
    p = {
        "material_path": material_path,
        "from_id": from_id, "to_id": to_id,
        "from_output": from_output, "to_input": to_input,
    }
    return _call_cpp(unreal.ArborMaterialGraphTools.connect_material_nodes(json.dumps(p)))


def connect_output(material_path, expression_id, property, *, from_output=""):
    """Wire an expression's output to a material output channel.

    `property` accepts: BaseColor, Normal, Roughness, Metallic, EmissiveColor,
    Opacity, OpacityMask, AmbientOcclusion, WorldPositionOffset, Refraction,
    PixelDepthOffset, SubsurfaceColor, Tangent, Anisotropy, Specular.
    On UE 5.7+ also: FrontMaterial, SurfaceThickness, Displacement.
    """
    p = {
        "material_path": material_path,
        "expression_id": expression_id,
        "property": property,
        "from_output": from_output,
    }
    return _call_cpp(unreal.ArborMaterialGraphTools.connect_material_output(json.dumps(p)))


def recompile_material(path):
    """Explicit terminal recompile + save. Call once after batched edits."""
    return _call_cpp(unreal.ArborMaterialGraphTools.recompile_material_asset(path))


def build_material(spec):
    """Build or update a complete material from a JSON spec. Idempotent.

    See ArborMaterialGraphTools.h::BuildMaterial for spec schema.
    """
    return _call_cpp(unreal.ArborMaterialGraphTools.build_material(json.dumps(spec)))


def layout_material(path):
    """Auto-arrange a material or material function's nodes into a readable layout.

    Detects whether `path` is a Material or a MaterialFunction and runs UE's
    built-in LayoutMaterialExpressions / LayoutMaterialFunctionExpressions, then
    saves. build_material / build_material_function already do this at the end
    unless the spec sets "auto_layout": False (which preserves manual x/y).
    """
    return _call_cpp(unreal.ArborMaterialGraphTools.layout_material(json.dumps({"path": path})))


def build_material_checked(spec):
    """Build a material AND verify it actually compiled, in one call.

    Builds the spec, forces a synchronous shader compile, then reads back the
    compile errors. Returns the build result augmented with:
        - compile_errors: list[str]  (empty = compiled clean)
        - all_wired:       bool       (every spec connection resolved)
        - ok:              bool       (success AND no compile errors AND wired)

    This collapses the build -> recompile -> query_material(compile_errors)
    loop you'd otherwise run by hand every time. A material that renders as the
    default lit-grey sphere has FAILED to compile - read `compile_errors` for
    the reason (e.g. "(Node ComponentMask) Missing ComponentMask input") instead
    of guessing from screenshots.
    """
    result = build_material(spec) or {}
    path = spec.get("path", "")
    _force_compile(path)
    q = query_material(path) if path else {}
    errors = q.get("compile_errors") or []
    expected = len(spec.get("connections") or [])
    actual = result.get("connection_count")
    result["compile_errors"] = errors
    result["all_wired"] = (actual == expected) if actual is not None else None
    result["ok"] = bool(result.get("success")) and not errors and result["all_wired"] is not False
    return result


# ---------------------------------------------------------------------------
# Material Function authoring (Phase 7) — delegates to ArborMaterialGraphTools
#
# Same sentinel-ID / reflection machinery as build_material, but the asset is a
# UMaterialFunction. A function has NO material-output pins or flags: its inputs
# are MaterialExpressionFunctionInput nodes and its outputs are
# MaterialExpressionFunctionOutput nodes. Wire math into a FunctionOutput's
# (unnamed) input pin via the normal connections list with an empty to_input.
# ---------------------------------------------------------------------------

def query_material_function(path):
    """Return a JSON snapshot of a Material Function's graph.

    Returns:
        dict with keys: success, function_path, description, expose_to_library,
        library_categories, expressions, connections, inputs, outputs.
    """
    return _call_cpp(unreal.ArborMaterialGraphTools.query_material_function(path))


def build_material_function(spec):
    """Build or update a complete UMaterialFunction from a JSON spec. Idempotent.

    See ArborMaterialGraphTools.h::BuildMaterialFunction for the spec schema.
    Minimal example (signed-distance circle):

        build_material_function({
            "path": "/Game/Assets/Materials/Functions/Procedural/MF_SDF_Circle",
            "description": "Signed distance to a circle; negative inside.",
            "expose_to_library": True,
            "library_categories": ["Procedural", "SDF"],
            "expressions": [
                {"id": "in_uv", "class": "MaterialExpressionFunctionInput",
                 "properties": {"InputName": "UV",
                                "InputType": "FunctionInput_Vector2",
                                "SortPriority": 0}},
                {"id": "out_d", "class": "MaterialExpressionFunctionOutput",
                 "properties": {"OutputName": "Distance", "SortPriority": 0}},
                # ... math nodes ...
            ],
            "connections": [
                {"from": "len", "to": "sub", "to_input": "A"},
                {"from": "sub", "to": "out_d", "to_input": ""},
            ],
        })

    Returns:
        dict with keys: success, function_path, expression_count,
        connection_count, inputs, outputs.
    """
    return _call_cpp(unreal.ArborMaterialGraphTools.build_material_function(json.dumps(spec)))


def render_thumbnail(material_path, output_path, width=256, height=256, ensure_compiled=True):
    """Render a material's thumbnail to a PNG file via the UE thumbnail pipeline.

    Args:
        material_path:   e.g. "/Game/Materials/M_Concrete"
        output_path:     absolute filesystem path; parent dirs auto-created
        width, height:   pixel dims (default 256x256)
        ensure_compiled: force shaders to compile first (default True) so a
                         freshly-built material doesn't capture as the default
                         lit-grey placeholder. Set False to skip (e.g. batch
                         rendering already-compiled assets).
    """
    if ensure_compiled:
        _force_compile(material_path)
    p = {"material_path": material_path, "output_path": output_path,
         "width": width, "height": height}
    return _call_cpp(unreal.ArborMaterialGraphTools.render_material_thumbnail(json.dumps(p)))
